pkg/controllers/interruption/controller.go

```diff
@@ -0,0 +1,275 @@
+	"github.com/aws/karpenter/pkg/controllers/providers"
+	"github.com/aws/karpenter/pkg/errors"
+	"github.com/aws/karpenter/pkg/utils"
+
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/metrics"
+	operatorcontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+func init() {
+	lo.Must0(apis.AddToScheme(scheme.Scheme))
+}
+
+// Controller is an AWS interruption controller.
+// It continually polls an provisioned SQS queue for events from aws.ec2 and aws.health that
+// trigger node health events or node spot interruption/rebalance events.
+type Controller struct {
+	kubeClient                client.Client
+	clk                       clock.Clock
+	cluster                   *state.Cluster
+	recorder                  events.Recorder
+	provider                  *providers.SQS
+	unavailableOfferingsCache *awscache.UnavailableOfferings
+	parser                    *EventParser
+}
+
+func NewController(kubeClient client.Client, clk clock.Clock, recorder events.Recorder, cluster *state.Cluster,
+	sqsProvider *providers.SQS, unavailableOfferingsCache *awscache.UnavailableOfferings) *Controller {
+
+	return &Controller{
+		kubeClient:                kubeClient,
+		clk:                       clk,
+		cluster:                   cluster,
+		recorder:                  recorder,
+		provider:                  sqsProvider,
+		unavailableOfferingsCache: unavailableOfferingsCache,
+		parser:                    NewEventParser(DefaultParsers...),
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	defer metrics.Measure(reconcileDuration)()
+	list := &v1alpha1.AWSNodeTemplateList{}
+	if err := c.kubeClient.List(ctx, list); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing node templates, %w", err)
+	}
+
+	if awssettings.FromContext(ctx).EnableInterruptionHandling && len(list.Items) > 0 {
+		active.Set(1)
+		sqsMessages, err := c.provider.GetSQSMessages(ctx)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("getting messages from queue, %w", err)
+		}
+		if len(sqsMessages) == 0 {
+			return reconcile.Result{}, nil
+		}
+		instanceIDMap := c.makeInstanceIDMap()
+		errs := make([]error, len(sqsMessages))
+		workqueue.ParallelizeUntil(ctx, 10, len(sqsMessages), func(i int) {
+			errs[i] = c.handleMessage(ctx, instanceIDMap, sqsMessages[i])
+		})
+		return reconcile.Result{}, multierr.Combine(errs...)
+	}
+	active.Set(0)
+	return reconcile.Result{RequeueAfter: time.Second * 10}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) operatorcontroller.Builder {
+	return operatorcontroller.NewSingletonManagedBy(m).
+		Named("interruption")
+}
+
+func (c *Controller) LivenessProbe(_ *http.Request) error {
+	return nil
+}
+
+// handleMessage gets the node names of the instances involved in the queue message and takes the
```

Comment:
> and

When I see this, I think https://en.wikipedia.org/wiki/Single-responsibility_principle